tests/python/pants_test/backend/jvm/tasks/jvm_compile/zinc/zinc_compile_integration_base.py

```diff
@@ -343,6 +343,51 @@ def test_zinc_logs_warnings_properly(self):
         for expected in expected_strings:
             self.assertIn(expected, pants_run.stdout_data)
 
+    def _test_zinc_reports_diagnostic_counts(self, reporting):
+        with self.temporary_workdir() as workdir:
+            target = "testprojects/src/scala/org/pantsbuild/testproject/compilation_warnings/unused_import_warning:unused_import"
+            with self.temporary_cachedir() as cachedir:
+                args = ['--compile-rsc-args=+["-S-Ywarn-unused:_"]'] + (
+                    ["--compile-rsc-report-diagnostic-counts"] if reporting else []
+                )
+                pants_run = self.run_test_compile(workdir, cachedir, target, extra_args=args,)
+                self.assert_success(pants_run)
+
+            expected_strings = [
+                "Reporting number of diagnostics for: ScalaLibrary(testprojects/src/scala/org/pantsbuild/testproject/compilation_warnings/unused_import_warning:unused_import)",
+                "Error: 0",
+                "Warning: 1",
+                "Information: 0",
+                "Hint: 0",
+            ]
+
+            for expected in expected_strings:
+                if reporting:
+                    self.assertIn(expected, pants_run.stdout_data)
+                else:
+                    self.assertNotIn(expected, pants_run.stdout_data)
+            run_info_path = os.path.join(workdir, "run-tracker", "latest", "info")
+            with open(run_info_path, "r") as run_info:
+
+                def is_target_data_line(line):
+                    return line.startswith("target_data: ")
+
+                target_data_line = next(filter(is_target_data_line, run_info))
+                assert target_data_line is not None
```

Comment:
```suggestion
                target_data_line = assert_single_element(filter(is_target_data_line, run_info))
```


If there are no lines, `next` will raise a `StopIteration` rather than returning `None`:
```
$ python3.6 -c "next(iter([]))"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
StopIteration
```

You'll also need to `from pants.util.collections import assert_single_element`